### PR TITLE
fix non-responsive Ctrl-C during disk-usage sorting

### PR DESCRIPTION
**Pull request by danfarino**
Ctrl-C wasn't being noticed during a disk-usage sorting, effectively meaning there was no way to cancel the operation except to wait it out.

Tested on Ubuntu 20.04 and an old MacOS 10.11.6 (El Capitan).

Thanks for a great program!
-Dan

### CODE REVIEW

**Comment by jarun** [comment]
We want to skip a check at every file. Can we do this when we encounter a directory instead?

Please share the performance data of current and new in cache-dropped state as well.

**Comment by jarun** [comment]
Unrelated to this PR, currently 1 thread processes a sub-directory tree under the directory where you do du. Say, at /, 1 thread processes each thread. The problem is - /bin pr /lib takes muck longer. See if we can do 1 dir (not the full tree) per thread. I guess if we try that, we will have to store the subtree's top directory info with the directories as the size should reflect for that dir visually.

File: src/nnn.c [hunk 1]
```diff
@@ -346,6 +346,7 @@ typedef struct {
 
 /* Non-persistent program-internal states (alphabeical order) */
 typedef struct {
+	uint_t interrupt;      	/* Program received an interrupt */
```
[the remark this answers is not in the document]
Comment (jarun) [reply src/nnn.c:349]: I don't think the threads need to read this atomically as none of them changes it.

**Review by jarun — CHANGES_REQUESTED** [review]

**Comment by danfarino** [comment]
Thanks for taking a look. I'm pretty new to nnn, and don't have the bandwidth to really dig in and learn the code in depth. That said, I saw that `g_state.interrupt` was being written to in the signal handler but those writes weren't being noticed in other parts of the program. Adding atomic operations fixed that issue on my machines. I had to make `interrupt` a `uint_t` instead of a bit field so that it would work with the atomic operations. (It's been years since I've done any serious C programming, so if I missed an obvious way to make it work without changing that, even better.)

I added the atomic operations around access to `active_threads` since I saw `while (active_threads == NUM_DU_THREADS);` and spinlocks definitely need atomic operations to work properly.

Please feel free to incorporate whatever ideas from this PR you'd like, and change the rest.

Thanks!

**Comment by jarun** [comment]
> That said, I saw that g_state.interrupt was being written to in the signal handler

That's global. Also we clear in global scope. Currently dirwalk doesn't add a new dir when the interrupt is set.

>  I'm pretty new to nnn, and don't have the bandwidth to really dig in and learn the code in depth. 

> Please feel free to incorporate whatever ideas from this PR you'd like, and change the rest.

I will refer to this PR and incorporate the changes when I get the time.

**Comment by jarun** [comment]
Commit 4c705a97fc0bb97b83babb7a3996939793582067 should work for you.

**Comment by danfarino** [comment]
Yes! Works great so far. Much simpler than what I had. Thanks!

**Comment by jarun** [comment]
Thanks for the confirmation.

**Pull request closed without merging**